File: api/lib/secretCrypto.js
import crypto from "node:crypto";

// Encrypts small user-supplied credentials (e.g. a user's own Apify API
// token) before they touch the database. Separate from JWT_SECRET/
// INTERNAL_SCRAPE_SECRET on purpose -- those authenticate our own tokens
// and are fine to rotate freely; this key decrypts data at rest, so
// rotating it silently breaks every previously-saved token.
//
// SECRET_ENCRYPTION_KEY must be 32 raw bytes, base64-encoded (openssl rand
// -base64 32). Stored ciphertext is "<iv>:<authTag>:<ciphertext>", each
// base64, so it's one TEXT column with no extra columns to keep in sync.

const ALGO = "aes-256-gcm";

function loadKey() {
  const raw = process.env.SECRET_ENCRYPTION_KEY;
  if (!raw) throw new Error("SECRET_ENCRYPTION_KEY is not configured");
  const key = Buffer.from(raw, "base64");
  if (key.length !== 32) {
    throw new Error("SECRET_ENCRYPTION_KEY must decode to exactly 32 bytes (openssl rand -base64 32)");
  }
  return key;
}

export function encryptSecret(plaintext) {
  if (typeof plaintext !== "string" || !plaintext) return null;
  const key = loadKey();
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv(ALGO, key, iv);
  const ciphertext = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()]);
  const authTag = cipher.getAuthTag();
  return [iv.toString("base64"), authTag.toString("base64"), ciphertext.toString("base64")].join(":");
}

export function decryptSecret(stored) {
  if (!stored) return null;
  const key = loadKey();
  const parts = stored.split(":");
  if (parts.length !== 3) return null;
  const [ivB64, tagB64, ctB64] = parts;
  try {
    const decipher = crypto.createDecipheriv(ALGO, key, Buffer.from(ivB64, "base64"));
    decipher.setAuthTag(Buffer.from(tagB64, "base64"));
    const plaintext = Buffer.concat([decipher.update(Buffer.from(ctB64, "base64")), decipher.final()]);
    return plaintext.toString("utf8");
  } catch {
    // Wrong key, corrupted row, or tampered ciphertext -- treat as absent
    // rather than crashing the scrape that was about to use it.
    return null;
  }
}

File: scraper/secret_crypto.py
"""
Python port of api/lib/secretCrypto.js's decryptSecret -- needed here
because the nightly/weekly cron scrapers talk to Postgres directly
(they're not going through the Node API), but still need to decrypt
each BOM owner's stored Apify token to use per-owner "bring your own
token" credits instead of a single shared APIFY_TOKEN secret.

Must stay in sync with secretCrypto.js: AES-256-GCM, key is
SECRET_ENCRYPTION_KEY (32 raw bytes, base64), stored ciphertext is
"<iv>:<authTag>:<ciphertext>", each base64.
"""

import os
import base64
from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def _load_key() -> bytes:
    raw = os.environ.get("SECRET_ENCRYPTION_KEY")
    if not raw:
        raise RuntimeError("SECRET_ENCRYPTION_KEY is not configured")
    key = base64.b64decode(raw)
    if len(key) != 32:
        raise RuntimeError("SECRET_ENCRYPTION_KEY must decode to exactly 32 bytes (openssl rand -base64 32)")
    return key


def decrypt_secret(stored: str):
    """Mirrors secretCrypto.js decryptSecret: returns the plaintext
    string, or None on any failure (missing key, wrong key, corrupted
    row, tampered ciphertext) -- same "treat as absent" behavior so a
    bad token for one user doesn't blow up the whole cron run.
    """
    if not stored:
        return None
    parts = stored.split(":")
    if len(parts) != 3:
        return None
    iv_b64, tag_b64, ct_b64 = parts
    try:
        key = _load_key()
        iv = base64.b64decode(iv_b64)
        tag = base64.b64decode(tag_b64)
        ciphertext = base64.b64decode(ct_b64)
        aesgcm = AESGCM(key)
        # Node's crypto keeps the GCM auth tag separate; the `cryptography`
        # lib expects it appended to the ciphertext.
        plaintext = aesgcm.decrypt(iv, ciphertext + tag, None)
        return plaintext.decode("utf-8")
    except Exception:
        return None
